File: src/timerapp_ag/webdav_config.py
"""Локальные настройки WebDAV (не попадают в синхронизируемый data.json)."""
from __future__ import annotations

import json
import os
from dataclasses import dataclass
from datetime import datetime, timedelta
from typing import Any
from urllib.parse import urljoin
from uuid import uuid4

from . import platform_paths
from .domain.datetime_util import local_now, parse_iso_datetime
from .secure_files import write_json_secrets
from .webdav_meta import META_SUFFIX

DEFAULT_REMOTE_PATH = "tasktimer/data.json"
REMIND_LATER_MINUTES_CHOICES = (5, 10, 15, 30, 60)
DEFAULT_REMIND_LATER_MINUTES = 15


@dataclass
class WebDavConfig:
    enabled: bool = False
    url: str = ""
    username: str = ""
    password: str = ""
    remote_path: str = DEFAULT_REMOTE_PATH
    sync_on_startup: bool = True
    sync_on_shutdown: bool = True
    shutdown_upload_only: bool = False
    sync_interval_minutes: int = 0
    sync_remind_later_minutes: int = DEFAULT_REMIND_LATER_MINUTES
    last_sync_at: str | None = None
    last_error: str = ""
    device_id: str = ""
    last_remote_content_hash: str = ""
    last_sync_had_conflict: bool = False
    pending_notice: str = ""
    pending_remote_hash: str = ""
    pending_remote_remind_at: str | None = None

    def to_dict(self) -> dict[str, Any]:
        return {
            "enabled": self.enabled,
            "url": self.url,
            "username": self.username,
            "password": self.password,
            "remote_path": self.remote_path,
            "sync_on_startup": self.sync_on_startup,
            "sync_on_shutdown": self.sync_on_shutdown,
            "shutdown_upload_only": self.shutdown_upload_only,
            "sync_interval_minutes": self.sync_interval_minutes,
            "sync_remind_later_minutes": self.sync_remind_later_minutes,
            "last_sync_at": self.last_sync_at,
            "last_error": self.last_error,
            "device_id": self.device_id,
            "last_remote_content_hash": self.last_remote_content_hash,
            "last_sync_had_conflict": self.last_sync_had_conflict,
            "pending_notice": self.pending_notice,
            "pending_remote_hash": self.pending_remote_hash,
            "pending_remote_remind_at": self.pending_remote_remind_at,
        }

    @classmethod
    def from_dict(cls, data: dict[str, Any] | None) -> WebDavConfig:
        if not isinstance(data, dict):
            return cls()
        return cls(
            enabled=bool(data.get("enabled", False)),
            url=str(data.get("url", "") or "").strip(),
            username=str(data.get("username", "") or "").strip(),
            password=str(data.get("password", "") or ""),
            remote_path=str(data.get("remote_path") or DEFAULT_REMOTE_PATH).strip() or DEFAULT_REMOTE_PATH,
            sync_on_startup=bool(data.get("sync_on_startup", True)),
            sync_on_shutdown=bool(data.get("sync_on_shutdown", True)),
            shutdown_upload_only=bool(data.get("shutdown_upload_only", False)),
            sync_interval_minutes=max(0, int(data.get("sync_interval_minutes") or 0)),
            sync_remind_later_minutes=normalize_remind_later_minutes(
                int(data.get("sync_remind_later_minutes") or DEFAULT_REMIND_LATER_MINUTES)
            ),
            last_sync_at=data.get("last_sync_at"),
            last_error=str(data.get("last_error", "") or ""),
            device_id=str(data.get("device_id") or "").strip(),
            last_remote_content_hash=str(data.get("last_remote_content_hash") or "").strip(),
            last_sync_had_conflict=bool(data.get("last_sync_had_conflict", False)),
            pending_notice=str(data.get("pending_notice") or "").strip(),
            pending_remote_hash=str(data.get("pending_remote_hash") or "").strip(),
            pending_remote_remind_at=data.get("pending_remote_remind_at"),
        )

    def is_configured(self) -> bool:
        return bool(self.url.strip() and self.username.strip())

    def remote_url(self) -> str:
        base = self.url.strip()
        if not base.endswith("/"):
            base += "/"
        path = self.remote_path.strip().lstrip("/")
        return urljoin(base, path)

    def meta_remote_url(self) -> str:
        data_url = self.remote_url()
        if data_url.endswith(".json"):
            return data_url[: -len(".json")] + META_SUFFIX
        return data_url.rstrip("/") + META_SUFFIX

    def ensure_device_id(self) -> str:
        if not self.device_id:
            self.device_id = uuid4().hex
        return self.device_id


def normalize_remind_later_minutes(value: int) -> int:
    if value in REMIND_LATER_MINUTES_CHOICES:
        return value
    return DEFAULT_REMIND_LATER_MINUTES


def should_show_remote_prompt(config: WebDavConfig, remote_hash: str) -> bool:
    """Показывать ли запрос на pull для данного hash (учитывает «Позже» и таймер повтора)."""
    pending = (config.pending_remote_hash or "").strip()
    if not pending or pending != remote_hash:
        return True
    remind_at_raw = (config.pending_remote_remind_at or "").strip()
    if not remind_at_raw:
        return True
    try:
        remind_at = parse_iso_datetime(remind_at_raw)
    except ValueError:
        return True
    return local_now() >= remind_at


def clear_pending_remote_remind(config: WebDavConfig) -> WebDavConfig:
    updated = WebDavConfig.from_dict(config.to_dict())
    updated.pending_remote_hash = ""
    updated.pending_remote_remind_at = None
    save_webdav_config(updated)
    return updated


def save_pending_remote_remind(config: WebDavConfig, remote_hash: str) -> WebDavConfig:
    minutes = normalize_remind_later_minutes(config.sync_remind_later_minutes)
    remind_at = local_now() + timedelta(minutes=minutes)
    updated = WebDavConfig.from_dict(config.to_dict())
    updated.pending_remote_hash = remote_hash.strip()
    updated.pending_remote_remind_at = remind_at.isoformat(timespec="seconds")
    save_webdav_config(updated)
    return updated


def prepare_remote_prompt(config: WebDavConfig, remote_hash: str) -> WebDavConfig:
    """Сбросить «Позже», если на сервере появилась новая версия."""
    pending = (config.pending_remote_hash or "").strip()
    if pending and pending != remote_hash.strip():
        return clear_pending_remote_remind(config)
    return config


def webdav_config_path():
    return platform_paths.webdav_config_path()


def apply_env_defaults(config: WebDavConfig, *, respect_saved_enabled: bool = False) -> WebDavConfig:
    """Подставить WEBDAV_* из окружения, если поля пустые."""
    url = (os.environ.get("WEBDAV_URL") or "").strip()
    username = (os.environ.get("WEBDAV_USERNAME") or os.environ.get("WEBDAV_USER") or "").strip()
    password = os.environ.get("WEBDAV_PASSWORD") or ""
    remote_path = (os.environ.get("WEBDAV_REMOTE_PATH") or "").strip()
    enabled_env = (os.environ.get("WEBDAV_ENABLED") or "").strip().lower()
    enabled_from_env = enabled_env in {"1", "true", "yes", "on"}
    if respect_saved_enabled:
        enabled = config.enabled
    else:
        enabled = config.enabled or enabled_from_env

    return WebDavConfig(
        enabled=enabled,
        url=config.url or url,
        username=config.username or username,
        password=config.password or password,
        remote_path=config.remote_path if config.remote_path != DEFAULT_REMOTE_PATH else (remote_path or DEFAULT_REMOTE_PATH),
        sync_on_startup=config.sync_on_startup,
        sync_on_shutdown=config.sync_on_shutdown,
        shutdown_upload_only=config.shutdown_upload_only,
        sync_interval_minutes=config.sync_interval_minutes,
        sync_remind_later_minutes=config.sync_remind_later_minutes,
        last_sync_at=config.last_sync_at,
        last_error=config.last_error,
        device_id=config.device_id,
        last_remote_content_hash=config.last_remote_content_hash,
        last_sync_had_conflict=config.last_sync_had_conflict,
        pending_notice=config.pending_notice,
        pending_remote_hash=config.pending_remote_hash,
        pending_remote_remind_at=config.pending_remote_remind_at,
    )


def load_webdav_config() -> WebDavConfig:
    path = platform_paths.webdav_config_path()
    if not path.is_file():
        config = apply_env_defaults(WebDavConfig())
        config.ensure_device_id()
        return config
    try:
        payload = json.loads(path.read_text(encoding="utf-8"))
    except (OSError, json.JSONDecodeError):
        config = apply_env_defaults(WebDavConfig())
        config.ensure_device_id()
        return config
    config = apply_env_defaults(WebDavConfig.from_dict(payload), respect_saved_enabled=True)
    config.ensure_device_id()
    return config


def save_webdav_config(config: WebDavConfig) -> None:
    config.ensure_device_id()
    write_json_secrets(platform_paths.webdav_config_path(), config.to_dict())


def mark_webdav_sync_ok(config: WebDavConfig, *, remote_hash: str = "", had_conflict: bool = False) -> WebDavConfig:
    updated = WebDavConfig.from_dict(config.to_dict())
    updated.last_sync_at = datetime.now().isoformat(timespec="seconds")
    updated.last_error = ""
    updated.last_sync_had_conflict = had_conflict
    if remote_hash:
        updated.last_remote_content_hash = remote_hash
    save_webdav_config(updated)
    return updated


def mark_webdav_sync_error(config: WebDavConfig, message: str) -> WebDavConfig:
    updated = WebDavConfig.from_dict(config.to_dict())
    updated.last_error = message.strip()
    save_webdav_config(updated)
    return updated


def save_webdav_pending_notice(message: str) -> None:
    config = load_webdav_config()
    updated = WebDavConfig.from_dict(config.to_dict())
    updated.pending_notice = message.strip()
    save_webdav_config(updated)


def peek_webdav_pending_notice() -> str:
    config = load_webdav_config()
    return (config.pending_notice or "").strip()


def clear_webdav_pending_notice() -> None:
    config = load_webdav_config()
    if not (config.pending_notice or "").strip():
        return
    updated = WebDavConfig.from_dict(config.to_dict())
    updated.pending_notice = ""
    save_webdav_config(updated)


def consume_webdav_pending_notice() -> str:
    notice = peek_webdav_pending_notice()
    if notice:
        clear_webdav_pending_notice()
    return notice

File: android/app/src/main/java/com/timerapp/linkb24/data/WebDavConfig.kt
package com.timerapp.linkb24.data

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import java.util.UUID
import kotlin.math.max
import com.timerapp.linkb24.webdav.DEFAULT_REMIND_LATER_MINUTES
import com.timerapp.linkb24.webdav.REMIND_LATER_MINUTES_CHOICES
import com.timerapp.linkb24.webdav.metaRemotePath

const val DEFAULT_WEBDAV_REMOTE_PATH = "tasktimer/data.json"

@Serializable
data class WebDavConfig(
    val enabled: Boolean = false,
    val url: String = "",
    val username: String = "",
    val password: String = "",
    @SerialName("remote_path") val remotePath: String = DEFAULT_WEBDAV_REMOTE_PATH,
    @SerialName("sync_on_startup") val syncOnStartup: Boolean = true,
    @SerialName("sync_on_shutdown") val syncOnShutdown: Boolean = true,
    @SerialName("shutdown_upload_only") val shutdownUploadOnly: Boolean = false,
    @SerialName("sync_interval_minutes") val syncIntervalMinutes: Int = 0,
    @SerialName("sync_remind_later_minutes") val syncRemindLaterMinutes: Int = DEFAULT_REMIND_LATER_MINUTES,
    @SerialName("last_sync_at") val lastSyncAt: String? = null,
    @SerialName("last_error") val lastError: String = "",
    @SerialName("device_id") val deviceId: String = "",
    @SerialName("last_remote_content_hash") val lastRemoteContentHash: String = "",
    @SerialName("last_sync_had_conflict") val lastSyncHadConflict: Boolean = false,
    @SerialName("pending_notice") val pendingNotice: String = "",
    @SerialName("pending_remote_hash") val pendingRemoteHash: String = "",
    @SerialName("pending_remote_remind_at") val pendingRemoteRemindAt: String? = null,
) {
    fun isConfigured(): Boolean = url.trim().isNotEmpty() && username.trim().isNotEmpty()

    fun remoteUrl(): String {
        val base = url.trim().trimEnd('/') + "/"
        val path = remotePath.trim().trimStart('/')
        return base + path
    }

    fun metaRemoteUrl(): String = metaRemotePath(remoteUrl())

    fun withDeviceId(): WebDavConfig {
        if (deviceId.isNotBlank()) {
            return this
        }
        return copy(deviceId = UUID.randomUUID().toString().replace("-", ""))
    }

    fun periodicSyncEnabled(): Boolean = enabled && syncIntervalMinutes > 0 && isConfigured()
}

const val MIN_ANDROID_PERIODIC_SYNC_MINUTES = 15

fun normalizeSyncIntervalMinutes(value: Int): Int {
    val clamped = value.coerceIn(0, 1440)
    if (clamped == 0) {
        return 0
    }
    return max(clamped, MIN_ANDROID_PERIODIC_SYNC_MINUTES)
}

fun validateWebDavConfig(config: WebDavConfig): String? {
    if (!config.enabled) {
        return null
    }
    if (config.url.trim().isEmpty()) {
        return "Укажите URL WebDAV"
    }
    if (config.username.trim().isEmpty()) {
        return "Укажите имя пользователя"
    }
    if (config.remotePath.trim().isEmpty()) {
        return "Укажите путь к файлу на сервере"
    }
    if (config.syncIntervalMinutes !in 0..1440) {
        return "Интервал синхронизации: от 0 до 1440 минут"
    }
    if (config.syncRemindLaterMinutes !in REMIND_LATER_MINUTES_CHOICES) {
        return "Интервал напоминания: 5, 10, 15, 30 или 60 минут"
    }
    return null
}
